zamba/models/config.py

```diff
@@ -495,22 +495,44 @@ def preprocess_labels(cls, values):
                     labels["site"], proportions=values["split_proportions"]
                 )
             else:
+                # otherwise randomly allocate
                 logger.info(
-                    "No 'site' column found so videos will be randomly allocated to splits."
+                    "No 'site' column found so videos for each species will be randomly allocated across splits using provided split proportions."
                 )
-                # otherwise randomly allocate
+
+                expected_splits = [k for k, v in values["split_proportions"].items() if v > 0]
                 random.seed(SPLIT_SEED)
-                labels["split"] = random.choices(
-                    list(values["split_proportions"].keys()),
-                    weights=list(values["split_proportions"].values()),
-                    k=len(labels),
-                )
 
+                # check we have at least as many videos per species as we have splits
+                # labels are OHE at this point
+                num_videos_per_species = labels.filter(regex="species_").sum().to_dict()
+                too_few = {
+                    k.split("species_", 1)[1]: v
+                    for k, v in num_videos_per_species.items()
+                    if v < len(expected_splits)
+                }
+
+                if len(too_few) > 0:
+                    raise ValueError(
+                        f"Not all species have enough videos to allocate into the following splits: {', '.join(expected_splits)}. A minimumm of {len(expected_splits)} videos per label is required. Found the following counts: {too_few}. Either remove these labels or add more videos."
+                    )
+
+                for c in labels.filter(regex="species_").columns:
+                    species_df = labels[labels[c] > 0]
+
+                    # within each species, seed splits by putting one video in each set and then allocate videos based on split proportions
+                    labels.loc[species_df.index, "split"] = expected_splits + random.choices(
+                        list(values["split_proportions"].keys()),
+                        weights=list(values["split_proportions"].values()),
+                        k=len(species_df) - len(expected_splits),
+                    )
```

Comment:
This has one weird edge case that is likely rare, so just worth filing an issue for:

```
v0.mp4, antelope
v1.mp4, antelope  # v1 assigned test by antelope grouping
v1.mp4, cow       # subsequently v1 assigned train by cow grouping
v2.mp4, antelope
v4.mp4, cow
v5.mp4, cow

# test set is now missing antelope